### PR TITLE
Improve how we wait for connected container in tests

### PR DESCRIPTION
**Pull request by alexvy86**
## Description

This consolidates the way we wait for a connected container in our tests. The new way allows the caller to specify that the promise should reject if the container closes while the caller is waiting for it to connect. This defaults to false for now to maintain backwards compatibility but will be changed in an upcoming breaking change since rejecting on container close is a more sensible default behavior.

It's now also possible to provide the same timeout options that we expose in our general function to wait for a Promise with a timeout, which have no effect on the current behavior if not provided.

Our tests have been updated to call `waitForContainerConnection` in a way that they'll see a rejected promise if the container closes unexpectedly. This prevents an issue we saw with e2e tests, where instead of throwing the error that caused the closure and terminating immediately, they time out because the waiting promise never resolves. Mocha then reports a timeout instead of the actual underlying problem. We saw this behavior while testing changes to better handle non-retryable network errors (e.g. the server presenting an invalid SSL certificate) in #13449.

## Reviewer guidance

This PR intends to maintain backwards compatibility. Once this is merged and integrated into `next` I'll open another one with a breaking change to remove the deprecated function and change the default value for one of the parameters in the new one.

### CODE REVIEW

File: packages/test/test-utils/src/timeoutUtils.ts [hunk 1]
```diff
@@ -25,9 +25,17 @@ export async function timeoutAwait<T = void>(
...
 }
 
-export async function ensureContainerConnected(container: Container): Promise<void> {
+export async function ensureContainerConnected(
+    container: Container,
+    timeoutOptions: TimeoutWithError = {},
+    failOnContainerClose: boolean = true): Promise<void> {
     if (!container.connected) {
-        return timeoutPromise((resolve) => container.once("connected", () => resolve()));
+        return timeoutPromise((resolve, reject) => {
```
Comment (alexvy86) [inline packages/test/test-utils/src/timeoutUtils.ts:33]: Some of the tests that were consolidated use a slightly different check: 

```typescript
if (container.connectionState !== ConnectionState.Connected) {
    ...
}
```

I'm assuming that one and `if (!container.connected)` are equivalent, just leveraging the capabilities of `Container` vs `IContainer`.

File: packages/test/test-utils/src/timeoutUtils.ts [hunk 2]
```diff
@@ -25,9 +25,17 @@ export async function timeoutAwait<T = void>(
...
+            container.once("connected", () => resolve());
```
Comment (alexvy86) [inline packages/test/test-utils/src/timeoutUtils.ts:34]: Some of the tests that were consolidated use a slightly different check: 

```typescript
if (container.connectionState !== ConnectionState.Connected) {
    ...
}
```

I'm assuming that one and `if (!container.connected)` are equivalent, just leveraging the capabilities of `Container` vs `IContainer`.

Comment (Abe27342) [reply packages/test/test-utils/src/timeoutUtils.ts:34]: yep.

```
    public get connectionState(): ConnectionState {
        return this.connectionStateHandler.connectionState;
    }

    public get connected(): boolean {
        return this.connectionStateHandler.connectionState === ConnectionState.Connected;
    }
```

File: packages/test/test-utils/src/timeoutUtils.ts [hunk 3]
```diff
@@ -25,9 +26,17 @@ export async function timeoutAwait<T = void>(
     return Promise.race([promise, timeoutPromise<T>(() => { }, timeoutOptions)]);
 }
 
-export async function ensureContainerConnected(container: Container): Promise<void> {
-    if (!container.connected) {
-        return timeoutPromise((resolve) => container.once("connected", () => resolve()));
+export async function ensureContainerConnected(
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/timeoutUtils.ts:29]: There is another function `waitForContainerConnection` in `TestSummaryUtils.ts` that does this but without the timeout. Maybe consolidate that too?

File: packages/test/test-utils/src/timeoutUtils.ts [hunk 4]
```diff
@@ -25,9 +26,17 @@ export async function timeoutAwait<T = void>(
...
+    container: IContainer,
+    timeoutOptions: TimeoutWithError = {},
+    failOnContainerClose: boolean = true): Promise<void> {
+    if (container.connectionState !== ConnectionState.Connected) {
+        return timeoutPromise((resolve, reject) => {
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/timeoutUtils.ts:34]: Just FYI, I have noticed that when I run the ODSP tests locally on codespaces a lot of tests time out because it doesn't get to connected state in 250ms (which is the default timeout in timeoutPromise). Not sure if that's an issue with my codespaces only but something to keep in mind if this happens in CI or something.

Comment (alexvy86) [reply packages/test/test-utils/src/timeoutUtils.ts:34]: I did get a bit concerned that it would be too short for some environments or particular tests, and was waiting to see if we started seeing frequent timeouts in CI. Maybe we make 1s the default specifically for `ensureContainerConnected`?

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 3 above)
Comment (alexvy86) [reply packages/test/test-utils/src/timeoutUtils.ts:29]: Thanks for pointing it out! I'll move everything to the new function.

File: packages/test/test-utils/src/timeoutUtils.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/test/test-utils/src/timeoutUtils.ts:34]: I am not sure what a good value would be. Maybe start with 1 second and it can be adjusted based on test results? I would add a comment to the function either way saying that the timeout is x and it should be adjusted if tests timeout due to container taking more than x to connect.

**Comment by alexvy86** [comment]
> Since this is a breaking change for the test-utils package, should we not be targeting `next`?

Possibly; I wanted to run it by the team during office hours because I'm not very clear which of our test-related packages are supposed to be more internal and for which ones do we make strong guarantees around breaking changes.

**Review by agarwal-navin — APPROVED** [review]

**Comment by alexvy86** [comment]
@agarwal-navin if you have a chance, I'd appreciate another look at this PR now that it's been refactored to maintain back-compat.

File: packages/test/test-end-to-end-tests/src/test/gc/gcAttachmentBlobs.spec.ts [hunk 5]
```diff
@@ -16,7 +16,7 @@ import { describeNoCompat, ITestDataObject } from "@fluidframework/test-version-
 import { getUrlFromItemId, MockDetachedBlobStorage } from "../mockDetachedBlobStorage";
 import { getGCStateFromSummary } from "./gcTestSummaryUtils";
 
-const ensureContainerConnectedWriteMode = async (container: Container) => {
+const waitForContainerConnectionWriteMode = async (container: Container) => {
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcAttachmentBlobs.spec.ts:19]: Not part of this PR but I think this should be a utility too. It's used in 2-3 tests.

File: packages/test/test-utils/src/containerUtils.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * Waits for the specified container to emit a 'connected' event.
+ *
+ * @deprecated Use waitForContainerConnection instead.
+ * Note that an upcoming release will change the default parameters on that function to:
+ * - failOnContainerClose = true
+ * - timeoutOptions.durationMs = 1s
+ */
+export async function ensureContainerConnected(container: Container): Promise<void> {
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/containerUtils.ts:18]: Is this used outside our repo? I don't think so but I guess its better to be on the cautious side and have this in place.

Maybe we need the ability to mark these packages as internal and not provide back-compat guarantees.

File: packages/test/test-utils/src/containerUtils.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
+/*!
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/containerUtils.ts:1]: nit: Maybe this file should be called connectionUtils.ts.

File: packages/test/test-utils/src/containerUtils.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    if (container.connectionState !== ConnectionState.Connected) {
+
+        const executor: PromiseExecutor = (resolve, reject) => {
+            container.once("connected", () => resolve())
+            if (failOnContainerClose) {
+                container.once("closed", (error) =>  reject(error));
+            }
+        };
+
+        return timeoutOptions === undefined ? new Promise(executor) : timeoutPromise(executor, timeoutOptions);
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/containerUtils.ts:57]: Are tests failing on making timeoutPromise the default?

File: packages/test/test-utils/src/containerUtils.ts (hunk 6 above)
Comment (alexvy86) [reply packages/test/test-utils/src/containerUtils.ts:18]: Yeah, I brought up this PR in office hours yesterday to see if we're making the same back-compat guarantees on our test-related packages and we agreed we do, thus this refactor.

And we also agreed that there's probably room for improvement, mainly along the lines of cleaning up the current API surface, and somehow being able to differentiate between things we export for general use and things we export but just for internal purposes.

But for now, back-compat it is. Once this merges and integrates into next, I'll open another PR to remove it in the next major version.

File: packages/test/test-utils/src/containerUtils.ts (hunk 7 above)
Comment (alexvy86) [reply packages/test/test-utils/src/containerUtils.ts:1]: I can see both approaches. I thought of it as utility functions related to the container (so that would get a container as parameter), and right now the only thing we have are functions related to the container's connection. Once we have more utilities one way or the other we can keep the filename or change it.

File: packages/test/test-utils/src/containerUtils.ts (hunk 8 above)
Comment (alexvy86) [reply packages/test/test-utils/src/containerUtils.ts:57]: They don't fail, but for full back-compat I needed to ensure that the utility function will never time out on its own in current call sites; mocha would still stop the test due to its own timeout, but if no `timeoutOptions` are provided then it should wait indefinitely, which `timeoutPromise` can't do (defaults to a 250ms timeout if we don't pass one in).

**Review by agarwal-navin — APPROVED** [review]
